File: src/main/java/com/saomc/saoui/screens/ingame/HealthStep.java
package com.saomc.saoui.screens.ingame;

import com.saomc.saoui.GLCore;
import com.saomc.saoui.social.StaticPlayerHelper;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;

@SideOnly(Side.CLIENT)
public enum HealthStep { // Could be loaded from file. Currently the HPBar loads it's own colors.

    VERY_LOW(0.1F, 0xBD0000FF),
    LOW(0.2F, 0xF40000FF),
    VERY_DAMAGED(0.3F, 0xF47800FF),
    DAMAGED(0.4F, 0xF4BD00FF),
    OKAY(0.5F, 0xEDEB38FF),
    GOOD(1.0F, 0x93F43EFF),
    CREATIVE(-1.0F, 0xB32DE3FF);

    private final float healthLimit;
    private final int color;

    HealthStep(float limit, int rgba) {
        healthLimit = limit;
        color = rgba;
    }

    public static HealthStep getStep(Minecraft mc, EntityLivingBase entity, float time) {
        return getStep(entity, StaticPlayerHelper.getHealth(mc, entity, time) / StaticPlayerHelper.getMaxHealth(entity));
    }

    public static HealthStep getStep(EntityLivingBase entity, double health) {
        return entity instanceof EntityPlayer && ((EntityPlayer) entity).capabilities.isCreativeMode ? CREATIVE : getStep(health);
    }

    public static HealthStep getStep(double health) {
        HealthStep step = first();
        while ((health > step.getLimit()) && (step.ordinal() + 1 < values().length)) step = step.next();
        return step == CREATIVE ? GOOD : step;
    }

    private static HealthStep first() {
        return values()[0];
    }

    private HealthStep next() {
        return values()[ordinal() + 1];
    }

    private float getLimit() {
        return healthLimit;
    }

    public final void glColor() {
        GLCore.glColorRGBA(color);
    }

}

File: src/main/java/com/saomc/saoui/screens/ingame/HealthStep.kt
/*
 * Copyright (C) 2016-2019 Arnaud 'Bluexin' Solé
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.saomc.saoui.screens.ingame

import com.saomc.saoui.GLCore
import com.saomc.saoui.social.StaticPlayerHelper
import net.minecraft.client.Minecraft
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

@SideOnly(Side.CLIENT)
enum class HealthStep constructor(private val limit: Float, var rgba: Int) {

    VERY_LOW(0.1f, 0xBD0000FF.toInt()),
    LOW(0.2f, 0xF40000FF.toInt()),
    VERY_DAMAGED(0.3f, 0xF47800FF.toInt()),
    DAMAGED(0.4f, 0xF4BD00FF.toInt()),
    OKAY(0.5f, 0xEDEB38FF.toInt()),
    GOOD(1.0f, 0x93F43EFF.toInt()),
    CREATIVE(-1.0f, 0xB32DE3FF.toInt());

    fun rgba() = rgba

    private operator fun next(): HealthStep {
        return values()[ordinal + 1]
    }

    fun glColor() {
        GLCore.color(rgba)
    }

    companion object {

        fun getStep(mc: Minecraft, entity: EntityLivingBase, time: Float): HealthStep {
            return getStep(entity, (StaticPlayerHelper.getHealth(mc, entity, time) / StaticPlayerHelper.getMaxHealth(entity)).toDouble())
        }

        fun getStep(entity: EntityLivingBase, health: Double): HealthStep {
            return if (entity is EntityPlayer && (entity.capabilities.isCreativeMode || entity.isSpectator)) CREATIVE else getStep(health)
        }

        fun getStep(health: Double): HealthStep {
            var step = first()
            while (health > step.limit && step.ordinal + 1 < values().size) step = step.next()
            return if (step == CREATIVE) GOOD else step
        }

        private fun first(): HealthStep {
            return values()[0]
        }
    }

}
